src/core/server/elasticsearch/cluster_client.ts

```diff
@@ -0,0 +1,186 @@
+   */
+  private isClosed = false;
+
+  constructor(private readonly config: ElasticsearchClientConfig, private readonly log: Logger) {
+    this.client = new Client(parseElasticsearchClientConfig(config, log));
+  }
+
+  /**
+   * Calls specified {@param endpoint} with provided {@param clientParams} on behalf of the
+   * Kibana internal user.
+   * @param endpoint String descriptor of the endpoint e.g. `cluster.getSettings` or `ping`.
+   * @param clientParams A dictionary of parameters that will be passed directly to the Elasticsearch JS client.
+   * @param options Options that affect the way we call the API and process the result.
+   */
+  public async callAsInternalUser(
+    endpoint: string,
+    clientParams: Record<string, unknown> = {},
+    options?: CallAPIOptions
+  ) {
+    this.assertIsNotClosed();
+
+    return await callAPI(this.client, endpoint, clientParams, options);
+  }
+
+  /**
+   * Closes the cluster client. After that client cannot be used and one should
+   * create a new client instance to be able to interact with Elasticsearch API.
+   */
+  public close() {
+    if (this.isClosed) {
+      return;
+    }
+
+    this.isClosed = true;
+    this.client.close();
+
+    if (this.scopedClient !== undefined) {
+      this.scopedClient.close();
+    }
+  }
+
+  /**
+   * Creates an instance of `ScopedClusterClient` based on the configuration the
+   * current cluster client that exposes additional `callAsCurrentUser` method
+   * scoped to the provided {@param req}. Consumers shouldn't worry about closing
+   * scoped client instances, these will be automatically closed as soon as the
+   * original cluster client isn't needed anymore and closed.
+   * @param req Request the `ScopedClusterClient` instance will be scoped to.
+   */
+  public asScoped(req: { headers?: Headers } = {}) {
+    // It'd have been quite expensive to create and configure client for every incoming
+    // request since it involves parsing of the config, reading of the SSL certificate and
+    // key files etc. Moreover scoped client needs two Elasticsearch JS clients at the same
+    // time: one to support `callAsInternalUser` and another one for `callAsCurrentUser`.
+    // To reduce that overhead we create one scoped client per cluster client and share it
+    // between all scoped client instances.
+    if (this.scopedClient === undefined) {
+      this.scopedClient = new Client(
+        parseElasticsearchClientConfig(this.config, this.log, {
+          auth: false,
+          ignoreCertAndKey: !this.config.ssl || !this.config.ssl.alwaysPresentCertificate,
+        })
+      );
+    }
+
+    const headers = req.headers
+      ? filterHeaders(req.headers, this.config.requestHeadersWhitelist)
+      : req.headers;
+
+    return new ScopedClusterClient(
+      (...args) => this.callAsInternalUser(...args),
+      async (...args) => {
+        this.assertIsNotClosed();
+        return await callAPI(this.scopedClient!, ...args);
```

Comment:
We can, I just didn't want to spend time and resources on parsing config (reading cert/key/ca files, parsing hosts etc.) for scoped client in case it's not used (in some places we don't need scoped client at all). In practice it may not be an issue, but it's hard to say right now and since it's internal implementation detail it'd be easy to change if we decide to do so.

Reply:
ok. I just noticed that we create not some many instances (namely 8) and thought this is a premature optimisation.